scripts/gen-req-full-stats.ts

```diff
@@ -0,0 +1,125 @@
+import {
+  onboardingDataCollection,
+  semestersCollection,
+  toggleableRequirementChoicesCollection,
+  overriddenFulfillmentChoicesCollection,
+  courseFulfillmentStats,
+} from './firebase-config';
+
+import computeGroupedRequirementFulfillmentReports from '../src/requirements/requirement-frontend-computation';
+import computeFulfillmentStats from '../src/requirements/fulfillment-stats';
+import { createAppOnboardingData } from '../src/user-data-converter';
+
+import '../src/requirements/decorated-requirements.json';
+
+// idRequirementFrequency is a hashmap where the key is the requirement ID and the value is
+// an array of maps. Each element in the array represents a slot in the requirement.
+// The map is a hashmap where the key is the course ID and the value is the frequency of the course
+// in the slot.
+let idRequirementFrequency = new Map<string, Map<number, number>[]>();
+/**
+ * Computes the requirement fulfillment statistics for all users. This is done by iterating through
+ * all the users and computing the computeGroupedRequirementFulfillmentReports for each user.
+ * This returns groupedRequirementFulfillmentReport which is then passed to computeFulfillmentStats.
+ * GroupedRequirementFulfillmentReport is a list of RequirementFulfillmentReport where each
+ * RequirementFulfillmentReport represents a requirement and a list of courses that fulfill the
+ * requirement. computeFulfillmentStats then computes the frequency of each course in each slot
+ * of the requirement and stores it in idRequirementFrequency.
+ * @param _callback
+ * @throws Error when computeGroupedRequirementFulfillmentReports fails to compute the fulfillment stats
+ */
+async function computeRequirementFullfillmentStatistics(_callback) {
+  const semQuerySnapshot = await semestersCollection.get();
+  semQuerySnapshot.forEach(async doc => {
+    // obtain the user's semesters, onboarding data, etc...
+    const semestersAndPlans = await doc.data();
+    const onboardingData = (await onboardingDataCollection.doc(doc.id).get()).data();
+    const toggleableRequirementChoices =
+      (await toggleableRequirementChoicesCollection.doc(doc.id).get()).data() ?? {};
+    const overriddenFulfillmentChoices =
+      (await overriddenFulfillmentChoicesCollection.doc(doc.id).get()).data() ?? {};
+
+    // if the user has obboarding data, semesters, etc...
+    if (
+      onboardingData !== undefined &&
+      semestersAndPlans.semesters !== undefined &&
+      toggleableRequirementChoices !== undefined &&
+      overriddenFulfillmentChoices !== undefined
```

Comment:
These conditions will always be true, no? Above, we set these to `{}` if they are `undefined`.